x/tally/keeper/endblock_test.go

```diff
@@ -10,25 +10,103 @@ import (
+		timeout           bool
+		expExitCode       uint32
+	}{
+		{
+			name:              "full single commit-reveal",
+			memo:              "YzJWamRYSmxaR0YwWVE9PQ==",
+			replicationFactor: 1,
+			numCommits:        1,
+			numReveals:        1,
+			timeout:           false,
+			expExitCode:       0,
+		},
+		{
+			name:              "full 5 commit-reveals",
+			memo:              "ZnVsbCA1IGNvbW1pdC1yZXZlYWxz",
+			replicationFactor: 5,
+			numCommits:        5,
+			numReveals:        5,
+			timeout:           false,
+			expExitCode:       0,
+		},
+		{
+			name:              "commit timeout",
+			memo:              "Y29tbWl0IHRpbWVvdXQ=",
+			replicationFactor: 2,
+			numCommits:        0,
+			numReveals:        0,
+			timeout:           true,
+			expExitCode:       keeper.TallyExitCodeNotEnoughCommits,
+		},
+		{
+			name:              "commit timeout with 1 commit",
+			memo:              "Y29tbWl0IHRpbWVvdXQgd2l0aCAxIGNvbW1pdA==",
+			replicationFactor: 2,
+			numCommits:        1,
+			numReveals:        0,
+			timeout:           true,
+			expExitCode:       keeper.TallyExitCodeNotEnoughCommits,
+		},
+		{
+			name:              "commit timeout with 2 commits",
+			memo:              "Y29tbWl0IHRpbWVvdXQgd2l0aCAyIGNvbW1pdHM=",
+			replicationFactor: 2,
+			numCommits:        1,
+			numReveals:        0,
+			timeout:           true,
+			expExitCode:       keeper.TallyExitCodeNotEnoughCommits,
+		},
+		{
+			name:              "reveal timeout",
+			memo:              "cmV2ZWFsIHRpbWVvdXQ=",
+			replicationFactor: 2,
+			numCommits:        2,
+			numReveals:        0,
+			timeout:           true,
+			expExitCode:       keeper.TallyExitCodeNotEnoughReveals,
+		},
+		{
+			name:              "reveal timeout with 2 reveals",
+			memo:              "cmV2ZWFsIHRpbWVvdXQgd2l0aCAyIHJldmVhbHM=",
+			replicationFactor: 3,
+			numCommits:        3,
+			numReveals:        2,
+			timeout:           true,
+			expExitCode:       keeper.TallyExitCodeNotEnoughReveals,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			drID := f.commitRevealDataRequest(t, tt.memo, tt.replicationFactor, tt.numCommits, tt.numReveals, tt.timeout)
 
-	err := f.tallyKeeper.ProcessTallies(f.Context(), f.coreContractAddr)
-	require.NoError(t, err)
+			err := f.tallyKeeper.EndBlock(f.Context())
+			require.NoError(t, err)
 
-	dataResult, err := f.batchingKeeper.GetLatestDataResult(f.Context(), drID)
-	require.NoError(t, err)
-	require.Equal(t, uint32(0), dataResult.ExitCode)
+			dataResult, err := f.batchingKeeper.GetLatestDataResult(f.Context(), drID)
+			require.NoError(t, err)
+			require.Equal(t, tt.expExitCode, dataResult.ExitCode)
 
-	dataResults, err := f.batchingKeeper.GetDataResults(f.Context(), false)
-	require.NoError(t, err)
-	require.Equal(t, *dataResult, dataResults[0])
+			dataResults, err := f.batchingKeeper.GetDataResults(f.Context(), false)
+			require.NoError(t, err)
+			require.Contains(t, dataResults, *dataResult)
```

Comment:
Was having trouble implementing checks for before & after balances of the executors here. Will look into it soon.

Reply:
HY and I discussed this, just need to alter tests to account for rewards going into pending withdrawl, and the data request poster getting the remaining escrow refunded.

Reply:
Quick 2 cents: I don’t think we should test whether the payout messages get handled correctly on the chain side. The contract unit tests can take care of that and for integration the test suite is probably a better candidate.

Reply:
That also works :)